raiden/transfer/mediated_transfer/initiator_manager.py

```diff
@@ -64,129 +68,192 @@ def events_for_cancel_current_route(transfer_description) -> List[Event]:
+            available_routes=available_routes,
+            transfer_description=transfer_description,
+            pseudo_random_generator=pseudo_random_generator,
+            block_number=block_number,
+        )
+
+        events.extend(cancel_events)
+        events.extend(sub_iteration.events)
+
+        if sub_iteration.new_state is None:
+            # Here we don't delete the initiator state, but instead let it live.
+            # It will be deleted when the lock expires. We do that so that we
+            # still have an initiator payment task around to process the
+            # LockExpired message that our partner will send us.
+            # https://github.com/raiden-network/raiden/issues/3146#issuecomment-447378046
+            return TransitionResult(payment_state, events)
+
+        new_transfer = sub_iteration.new_state.transfer
+        payment_state.initiator_transfers[new_transfer.lock.secrethash] = sub_iteration.new_state
+
+    iteration = TransitionResult(payment_state, events)
+
+    return iteration
+
+
+def subdispatch_to_initiatortransfer(
+        payment_state: InitiatorPaymentState,
+        initiator_state: InitiatorTransferState,
+        state_change: StateChange,
         channelidentifiers_to_channels: ChannelMap,
         pseudo_random_generator: random.Random,
+        block_number: BlockNumber,
 ) -> TransitionResult:
-    channel_identifier = payment_state.initiator.channel_identifier
+    channel_identifier = initiator_state.channel_identifier
     channel_state = channelidentifiers_to_channels.get(channel_identifier)
     if not channel_state:
-        return TransitionResult(payment_state, list())
+        return TransitionResult(initiator_state, list())
 
-    sub_iteration = initiator.handle_block(
-        initiator_state=payment_state.initiator,
+    sub_iteration = initiator.state_transition(
+        initiator_state=initiator_state,
         state_change=state_change,
         channel_state=channel_state,
         pseudo_random_generator=pseudo_random_generator,
+        block_number=block_number,
     )
-    iteration = iteration_from_sub(payment_state, sub_iteration)
-    return iteration
 
+    if sub_iteration.new_state is None:
+        del payment_state.initiator_transfers[initiator_state.transfer.lock.secrethash]
 
-def handle_init(
+    return sub_iteration
+
+
+def subdispatch_to_all_initiatortransfer(
         payment_state: InitiatorPaymentState,
-        state_change: ActionInitInitiator,
+        state_change: StateChange,
         channelidentifiers_to_channels: ChannelMap,
         pseudo_random_generator: random.Random,
         block_number: BlockNumber,
 ) -> TransitionResult:
-    events: List[Event]
-    if payment_state is None:
-        sub_iteration = initiator.try_new_route(
-            old_initiator_state=None,
+    events = list()
+    for secrethash in list(payment_state.initiator_transfers.keys()):
```

Comment:
We need a way to tell other why we are copying the keys. At first I thought this was unnecessary.

Reply:
this is not dispatching to the cancelled routes, so the remove lock expire will never be sent

Reply:
There's no filter for the transfers i am iterating, which means that all transfers in the list will receive the state change. Unless i am missing something.

Reply:
My bad, I had the impression the transfers were removed from the mapping when they got cancelled.